fix: Fix expression box focus issues - used useEffect to isolate expr…

pleacu opened the pull request:
…ession box.

Still issues with caret positioning - will address later.
Fixes: #1957

riccardo-forina commented on ui-react/packages/atlasmap-core/src/services/mapping-management.service.ts:
```diff
@@ -31,6 +31,7 @@ import { MappingSerializer } from '../utils/mapping-serializer';
 import { DataMapperUtil } from '../common/data-mapper-util';
 import { PaddingField } from '../models/document-definition.model';
 import { LookupTableUtil } from '../utils/lookup-table-util';
+import { MappingUtil } from 'utils';
```
Better use relative paths. Unfortunately TS paths are not respected when building the package as an npm module (I found this the hard way).

pleacu replied:
actrually didn't need MappingUtil - good catch.

riccardo-forina commented on ui-react/packages/atlasmap/src/Atlasmap/utils/ui.ts:
```diff
@@ -160,8 +160,8 @@ export function fromMappingDefinitionToIMappings(
   isSource: boolean,
-): void {
-  initializationService.cfg.mappingService.executeFieldSearch(
+): any[] {
```
Please never use `any`, it removes the benefit of using Typescript in first place. If we don't need the return of the `executeFieldSearch` method then we shouldn't return it.

pleacu replied:
We do need the return of executeFieldSearch - It sends back an array of { displayName and a field node}.   I'm looking to see if I can send back just an array of strings.

riccardo-forina commented on ui-react/packages/atlasmap/src/UI/ConditionalExpressionInput.tsx:
```diff
@@ -1,30 +1,82 @@
+  onToggleExpressionMode,
+  trailerId,
+}) => {
+  const [, setCondExprEnabled] = useState<boolean>();
```
Sorry, this usage of `useState` is wrong. It doesn't make sense in React to have a state and then never using it. Is this some way to force the refresh of the child components? If that's the case, then the child components should be refactored to follow React's flow.

pleacu replied:
agree - it wasnt' needed

pleacu replied:
ended up needing this - see updates

riccardo-forina commented on ui-react/packages/atlasmap/src/UI/ConditionalExpressionInput.tsx:
```diff
@@ -1,30 +1,112 @@
+  onToggleExpressionMode,
+  trailerId,
+}) => {
+  const [conditionExpressionEnabled, setCondExprEnabled] = useState<boolean>(
```
If I follow the trail I see that the callback will eventually [set this flag](https://github.com/atlasmap/atlasmap/blob/ea7f43e4abec4684cc83718102b6b17af16adada/ui-react/packages/atlasmap-core/src/services/mapping-management.service.ts#L592). With this useState we are creating a copy of that flag. Since we want to have a single source of truth and since the expression box engine is provided by the RxJS code, we should make that flag available through the `AtlasmapProvider` and feed its value to this component.

pleacu replied:
is this issue significant enough to block merging or can this be addressed is a separate issue?  There are a few things left to be fixed (d&d, caret positioning, etc).  I'd like to get the bulk checked in before it goes stale.

pleacu replied:
Ok - I've addressed this issue with the latest change.

riccardo-forina commented on ui-react/packages/atlasmap/src/UI/ExpressionContent.tsx:
```diff
@@ -1,51 +1,27 @@
 import React from "react";
 import { Tooltip, Form, FormGroup } from "@patternfly/react-core";
-import { Observable } from "rxjs";
+import { Subscription } from "rxjs";
```
Picking this line but commenting on the whole file. This file is heavily coupled with the RxJS store, which is bad because it makes testing this component from a pure UI perspective very difficult. 

Since this component is in the `UI` folder it shouldn't really know or care about how things work under the hood. It should care about how things are displayed to the user and how the user can interact with the component, while exposing the necessary callbacks for the consumer of this component to do glue it to the business logic.

An handy way to understand if a component is too coupled is writing a test and/or a story for it. If you have to bring into the story too many bits it's a signal that the component is too coupled.

pleacu replied:
What's evident about this file is that it has little to do with the UI - it's just a DIV in a react form that calls a typeahead search select.  It is likely difficult to test - it was difficult to write - but it doesn't violate any separation of duties set out in the architecture and it is providing a very specific purpose.  I think it's isolated well and attempting to refactor it into multiple abstractions to satisfy testability is wrong.

riccardo-forina replied:
Then the UI component should be thin and expose the callbacks needed to wire it to the code that knows about the logic. Think about how e.g. the `select` html component works, it's generic and exposes callbacks you can use to do your own specialized work.

If this UI component should only show some text and open a drawer when a special key (`@`) is keyed down, it should do only that.

All the rest should stay away, in the `Atlasmap` folder I'd say. And IMHO there should be no need for these function to know about the input field or the html that's been used to render the widget. If the interface between the logic and the ui uses basic types like `string`, then they are decoupled from each other allowing us to change the UI in the future without worrying about breaking the logic.

riccardo-forina commented on ui-react/packages/atlasmap/src/UI/ConditionalExpressionInput.tsx:
```diff
@@ -1,30 +1,106 @@
+  onToggleExpressionMode,
+  trailerId,
+}) => {
+  function onToggle() {
```
This function does nothing useful, it's wrapping a callback without adding any value to it. Also the callback is already called at line #99, so there are two ways to do the same thing.

pleacu replied:
yep - I've removed onToggleExpressionMode from the EclipseContent props

riccardo-forina commented on ui-react/packages/atlasmap/src/UI/ConditionalExpressionInput.tsx:
```diff
@@ -1,30 +1,106 @@
+            aria-label="Enable/ Disable conditional mapping expression"
+            tabIndex={-1}
+            onClick={onToggle}
+            disabled={!onConditionalMappingExpressionEnabled()}
```
From this component standpoint, a `boolean` is enough to know if the component needs to be disabled or not. If calling a function to get the boolean is required, it should be done on the caller side.

pleacu replied:
The boolean that controls whether its disabled is part of the mapping.  It needs a function call - this is the correct way to do this.

riccardo-forina replied:
`onConditionalMappingExpressionEnabled` is being called twice at each render. One here, the second time [here](https://github.com/atlasmap/atlasmap/pull/1982/files/3b94234a515c59699149785e94a3e251e302f0c4#diff-e5c48eedbcf5dc5fab5b849bda1e04e4R88). 

Ultimately, [this code](https://github.com/atlasmap/atlasmap/blob/d7f50903a3535a6cb8dfea91c6e3a6dbe74e6219/ui-react/packages/atlasmap-core/src/services/mapping-management.service.ts#L569) is being called, which does nothing but returning the value of a boolean.

I stand my point, this doesn't have to be a function because. Even more so, since it returns a value that deeply nested into the state used by RxJS, it's completely opaque to React's lifecycle. A change on that boolean will not trigger a render on the React side and will lead to a stale UI and bugs. I think right now it works only because we are registered to a number of observables on the `AtlasmapProvider` so that we "catch all changes" so to speak.

riccardo-forina commented on ui-react/packages/atlasmap/src/UI/ConditionalExpressionInput.tsx:
```diff
@@ -1,30 +1,106 @@
+    nodeId?: string,
+    startOffset?: number,
+    endOffset?: number,
+  ) => any;
```
Never use `any`

pleacu replied:
will address in a new issue

riccardo-forina commented on ui-react/packages/atlasmap/src/UI/ConditionalExpressionInput.tsx:
```diff
@@ -1,30 +1,106 @@
+    nodeId?: string | undefined,
+    offset?: number | undefined,
+  ) => void;
+  mappingExpressionObservable: () => any;
```
Never use `any`

riccardo-forina commented on ui-react/packages/atlasmap/src/UI/ConditionalExpressionInput.tsx:
```diff
@@ -1,30 +1,106 @@
 const styles = StyleSheet.create({
   toolbarItem: { flex: 1 },
 });
+export interface IConditionalExpressionInputProps {
```
This interface should extend `IExpressionContentProps` to avoid re-declaring the same stuff twice.

pleacu replied:
will address in a new issue

pleacu commented:
@riccardo-forina - this addresses all issues except the mappingExpressionClearText use of any.  Would still like to work on that as a separate issue to get exposure for what's here.

pleacu commented:
Ok - will merge and generate 2 issues.  One for any 'any' referenced in ExpressionComntent and one for making onConditionalMappingExpressionEnabled() a reference to the actual boolean instead of a fcn.
thkx for your patience...
